### PR TITLE
Cortex-M: fix quantized lowering for decomposed attention

### PR DESCRIPTION
**Pull request by amacharla15**
## Summary

Fix Cortex-M lowering for explicitly decomposed Transformer attention so that quantized `bmm`, softmax, and scalar-multiply operations can lower to Cortex-M operators.

This does not add general Transformer support or automatically decompose Transformer operators. The regression test explicitly applies the existing SDPA and LayerNorm decomposition passes before PT2E quantization.

## Problem

After decomposing Transformer attention before quantization, `_softmax` and `bmm` could be quantized, but the attention scaling `mul` did not lower correctly.

There were three related issues in the Cortex-M lowering pipeline:

1. Scalar multiplication is represented as an `aten.full` tensor after scalar-to-tensor replacement. For attention scaling, this rank-1 scalar tensor needs to be materialized at the MUL output shape before quantization so that Cortex-M quantized MUL can be selected.

2. After `FoldAndAnnotateQParamsPass`, output requantization information is stored in node metadata while the graph still contains `aten.linear`. Generic ATen fake/meta retracing therefore reconstructs the linear output as `int32`, even though the Cortex-M quantized linear output is requantized to `int8`. This later causes an `int8`/`int32` mismatch when retracing BMM.

3. Quantized scalar constants needed by MUL must be materialized after qparams are attached, without globally folding unrelated constant operations.

## Changes

- Add Cortex-M-specific retracing behavior so folded quantized `aten.linear` nodes use the dtype described by their output qparams.
- Use this retracing behavior in scalar replacement and quant-node replacement passes.
- Materialize only scalar-generated `aten.full` operands when their shape must match a MUL output.
- Leave general tensor broadcasting unchanged so existing fallback behavior is preserved.
- Materialize only quantized `aten.full` constants that eventually feed a quantized MUL.
- Add a regression test using a small `torch.nn.Transformer` with explicit SDPA and LayerNorm decomposition.

The regression verifies that decomposed attention lowers to:

- `cortex_m.quantized_batch_matmul`
- `cortex_m.softmax`
- `cortex_m.quantized_mul`

and that decomposed attention no longer retains the corresponding ATen BMM or softmax operations.

## Validation

Local validation on the updated implementation:

- Transformer regression passes with numerical comparison at `qtol=2`.
- Cortex-M MUL dialect tests: `9 passed, 4 xfailed`.
- Silero VAD regression passes.
- Shared-qspec regression preserves the expected xfail.
- Full Cortex-M dialect suite: `240 passed, 1 skipped, 24 xfailed`.
- `lintrunner -a` passes, including mypy.

### Cortex-M55 FVP validation

As additional validation, I serialized the decomposed Transformer and tested it end-to-end on the Cortex-M55 Corstone-300 FVP.

The stock Cortex-M runner initially could not load the full decomposed Transformer because its selective operator list does not include several portable operators present in the resulting program:

- `aten::permute_copy.out`
- `aten::squeeze_copy.dims_out`
- `aten::expand_copy.out`
- `aten::mean.out`
- `aten::var.correction_out`

I also checked the remaining `expand_copy` operations explicitly. The decomposed Edge graph contains 12 `aten.expand_copy` nodes before the Cortex-M backend passes run, and the lowered graph contains the same 12 afterward. The narrowed `MaterializeMulBroadcastPass` in this PR therefore does not introduce additional `expand_copy` operations; these shape operations are already present before Cortex-M lowering.

For local runtime validation only, I rebuilt the Cortex-M runner with those five existing portable kernels enabled. No PR source files were changed, and the working tree remained clean.

With that runner, the full path completed successfully:

- PT2E quantization
- Cortex-M lowering
- ExecuTorch serialization
- Cortex-M55 Corstone-300 FVP execution
- numerical output comparison with `qtol=2`

The end-to-end FVP run passed.

## Scope / limitations

This PR intentionally does not:

- add automatic Transformer decomposition to the Cortex-M backend,
- claim general or official Transformer support,
- force every decomposed Transformer MUL to use a Cortex-M quantized kernel,
- add general quantized tensor broadcasting support,
- or change the stock Cortex-M runner's selected portable operator set.

The changes are limited to making the existing explicit decomposition workflow lower the targeted quantized attention operations correctly while preserving existing behavior for unrelated Cortex-M graphs.

* The experimental `use_explicit_layout=True` pipeline is not covered by this fix. I reproduced the same folded `aten.linear` int8/int32 retracing failure there. Mirroring the legacy-path fixes gets past the initial failure, but the same issue then appears while retracing through `ReplaceSqueezeAndUnsqueezeWithViewPass` and `CortexMCanonicalizeViewCopyPermutePass`. Supporting decomposed Transformer attention in explicit-layout mode therefore requires a broader Cortex-M-aware retracing change.

### Additional validation after review fixes

* Cortex-M test suite, excluding FVP/implementation tests and the MobileNet v2, MobileNet v3, and Wav2Letter model files whose optional `torchvision`/`torchaudio` dependencies are not installed locally: `466 passed, 5 skipped, 36 xfailed`.
* Focused decomposed-SDPA regression passes numerical comparison at `qtol=2`.
* The focused SDPA model also serializes and executes end-to-end on the stock Cortex-M55 Corstone-300 FVP runner at `qtol=2`.
* `lintrunner -a` passes, including MYPY.

Addresses #21943

### CODE REVIEW

**Comment by amacharla15** [comment]
Thanks for triggering the Cortex-M CI.

The first CI run exposed a few regressions from the initial implementation that were broader than the intended fix:

- the MUL broadcast pass inserted `expand_copy` for existing dynamic broadcast cases, which caused Cortex-M runtime implementation tests to fail because that operator was not registered by the runner;
- running the generic constant-materialization pass globally also changed unrelated graphs, including folding the `arange` in the Silero VAD test and changing an existing shared-qspec expected-xfail case.

I narrowed both parts in the latest commit:

- `MaterializeMulBroadcastPass` now only materializes the scalar-generated `aten.full` operand needed for the decomposed-attention MUL and leaves general tensor broadcasting unchanged;
- constant materialization is now limited to quantized `aten.full` constants that feed quantized MUL instead of running the generic constant pass over the whole graph.

I also narrowed the Transformer regression so it checks the relevant decomposed attention lowering without requiring unrelated LayerNorm MULs to lower.

After these changes, locally:

- Transformer regression: `1 passed`
- Cortex-M MUL dialect tests: `9 passed, 4 xfailed`
- Silero VAD regression: `1 passed`
- shared-qspec case: restored to its expected `xfail`
- full Cortex-M dialect suite: `240 passed, 1 skipped, 24 xfailed`
- `lintrunner -a`: all checks passed

I updated the PR description to reflect the narrowed implementation and scope.

While the new CI run is waiting/running, I am also checking serialization/FVP loading of the lowered Transformer as an additional validation step.

**Comment by amacharla15** [comment]
Additional runtime validation on the latest commit (`e1171a6`):

I serialized the decomposed Transformer and tested it on the Cortex-M55 Corstone-300 FVP.

The stock Cortex-M runner initially could not load the full program because its selective operator list omits five portable operators present in the decomposed Transformer: `aten::permute_copy.out`, `aten::squeeze_copy.dims_out`, `aten::expand_copy.out`, `aten::mean.out`, and `aten::var.correction_out`.

The narrowed `MaterializeMulBroadcastPass` in this PR does not introduce `expand_copy`; it materializes the scalar `aten.full` directly at the MUL output shape. The remaining `expand_copy` nodes come from the broader decomposed Transformer graph.

For local validation only, I rebuilt the runner with those five existing portable kernels enabled. No PR source files were changed, and the working tree remained clean.

With that runner, the full path passed:

- PT2E quantization
- Cortex-M lowering
- ExecuTorch serialization
- Cortex-M55 Corstone-300 FVP execution
- numerical comparison with `qtol=2`

So the end-to-end FVP validation passes with the required portable operators registered.

The CI checks for `e1171a6` are currently running.

**Comment by amacharla15** [comment]
I also ran a few targeted ablations on the latest implementation:

* Without `MaterializeMulBroadcastPass`, the attention MUL no longer lowers to `cortex_m.quantized_mul`.
* Without `MaterializeQuantizedMulConstantsPass`, lowering still happens but numerical comparison fails (mean abs error ~0.107).
* Reverting either Cortex-M-specific retracing site reproduces the original `int8`/`int32` BMM mismatch.

So the two materialization passes are independently needed at different stages, and the retracing fix is required in both passes.

These were local-only checks; the working tree was restored afterward.

File: backends/cortex_m/passes/cortex_m_pass_manager.py [hunk 1]
```diff
@@ -53,7 +58,9 @@ class CortexMPassManager(PassManager):
         # Run before folding so qparams attach to max_pool2d values, not tuple + getitem.
         RemoveGetItemPass,
         FoldAndAnnotateQParamsPass,
-        ReplaceScalarWithTensorArgPass,
+        # Materialize only quantized MUL constants after qparams are attached.
+        MaterializeQuantizedMulConstantsPass,
+        CortexMReplaceScalarWithTensorArgPass,
```
[the remark this answers is not in the document]
Comment (amacharla15) [reply backends/cortex_m/passes/cortex_m_pass_manager.py:63]: I checked this path directly.

The decomposed Transformer reproduces the same `int8`/`int32` BMM failure with `use_explicit_layout=True`. I also tested mirroring `MaterializeQuantizedMulConstantsPass` and the Cortex-M scalar retracer into the explicit-layout pass list.

That gets past the initial `ReplaceScalarWithTensorArgPass` failure, but the same folded `aten.linear` metadata mismatch then appears in later explicit-layout retracing passes, including `ReplaceSqueezeAndUnsqueezeWithViewPass` and `CortexMCanonicalizeViewCopyPermutePass`.

So supporting this path requires broader Cortex-M-aware retracing across the explicit-layout pipeline rather than just mirroring these two passes. I kept this PR scoped to the legacy pipeline exercised by #21943 and documented the explicit-layout gap as a limitation.

File: backends/cortex_m/passes/materialize_mul_broadcast_pass.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            output_shape = list(output_tensor.shape)
+            new_args = list(node.args)
+
+            for index, arg in enumerate((lhs, rhs)):
+                tensor = get_first_fake_tensor(arg)
+                if tensor.shape == output_tensor.shape or not self._is_scalar_full(arg):
+                    continue
+
+                full_args = list(arg.args)
+                full_args[0] = output_shape
```
[the remark this answers is not in the document]
Comment (amacharla15) [reply backends/cortex_m/passes/materialize_mul_broadcast_pass.py:64]: I checked this as well.

Dynamic shapes are outside the supported Cortex-M backend contract. I tested the defensive option of skipping symbolic output shapes, but that would only avoid constructing the invalid FX node while leaving an unsupported broadcast graph in the pipeline.

Since this PR targets the existing static-shape Cortex-M workflow, I reverted that partial handling rather than implying dynamic-shape support. The scalar materialization here therefore remains limited to the supported static-shape path.

File: backends/cortex_m/passes/cortex_m_pass_manager.py [hunk 3]
```diff
@@ -41,9 +41,14 @@
     CortexMReplaceOpsWithChannelsLastVariants,
     ValidateCortexMExplicitLayoutPass,
 )
+from .materialize_mul_broadcast_pass import MaterializeMulBroadcastPass
+from .materialize_quantized_mul_constants_pass import (
+    MaterializeQuantizedMulConstantsPass,
+)
```
[the remark this answers is not in the document]
Comment (amacharla15) [reply backends/cortex_m/passes/cortex_m_pass_manager.py:47]: The new pass modules were available to source-based tests but were missing from the Buck targets.

I added `cortex_m_quantized_pass` as its own Python library target, declared it as a dependency of `replace_quant_nodes_pass` and `cortex_passes`, and added the new pass modules to `cortex_passes.srcs`.

I don't have `buck2` available in my local checkout, but the focused Cortex-M regressions still pass (`4 passed`) and the commit-time lintrunner completed successfully.

**Comment by amacharla15** [comment]
@zingo the workflows on `c09864c` are still awaiting approval, so CI hasn’t run on the current head yet. Could you approve them when you get a chance?

**Review by AdrianLundell — CHANGES_REQUESTED** [review]
Hello @amacharla15 and welcome with your first contribution to Executorch!

There are a number of things going on in this patch, as a general tip I strongly recommend splitting it up into several PR:s with one pass each for example, with briefer commit messages. This will significantly increase the chance of getting the changes merged  since it is much faster to review and an issue in one pass does not block all other passes for example.

In regards to the patch: 
Linear Q/DQ change: I happen to know that there will be a prototype for changing how the folded q/dq parameters are handled in the graph coming up just this week, so it makes sense to wait for that before merging  a new solution, bad timing unfortunately.

Matrialize mul change: Introducing full shape constants for a single scalar mul is not so efficient,  would prefer to see this by modifying the qparams, or a new accelerated kernel if there are cases where that does not work.

Matrialize constant change: Likely not need with other mul approach.

Let me know if you have any questions! Best regards,
Adrian

**Comment by amacharla15** [comment]
Thanks @AdrianLundell  ; 

Here is my observation and work: 

I split the linear and scalar-MUL issues apart and followed up on your suggestion to handle the scalar scale through qparams rather than materializing a full-shape constant.

While investigating that path, I found the existing draft PR #21027, which removes a `DQ -> MUL -> Q` sequence when the surrounding qparams already make the MUL an integer-preserving identity. I tested that approach against the decomposed `nn.Transformer` reproducer from #21943.

The SDPA decomposition splits the attention scale across Q and K, with each branch using `head_dim^(-1/4)`. On the full Transformer there are 6 such attention-scale branches. With independently observed qparams, only 2/6 satisfy the relation required by the fold:

`S_out ~= S_in * c` and `Z_out == Z_in`

The other 4 are correctly rejected by the fold, so those remaining MULs originate from the qparam choices made earlier during quantization rather than from the folding logic itself.

I then prototyped a small Cortex-M quantizer stage that derives the scaled activation qparams from the input:

`S_out = S_in * c` and `Z_out = Z_in`

for the positive scalar-scaled BMM-input pattern emitted by the existing SDPA decomposition.

One detail from the real graph: the scaled value reaches BMM through `expand` / `view`, so the derived qspec goes on the qspec-owning edge from `SharedQspecQuantizer` and the BMM input shares it. The stage only replaces an independently specified `QuantizationSpec` — existing shared or derived relationships are untouched.

With that change, all 6/6 attention-scale branches in the Transformer reproducer satisfy the identity relation needed by the later fold. I also tested the actual decomposed SDPA path with head dimensions 16 and 64, including the non-exact `64^(-1/4)` scale, and verified that the derived qparams reach both score-BMM operands.

I have kept the linear int8/int32 metadata issue separate and stopped pursuing the Cortex-M-specific retracing workaround there, since you mentioned the upcoming folded-Q/DQ prototype.

Does this derived-qparam approach look like the direction you had in mind for the scalar-MUL issue? If so, I can clean it up and send it as a separate focused PR.

**Comment by AdrianLundell** [comment]
Sounds like good progress, I like this direction so definitely worth opening a PR!

As a general heads up I prefer to keep the quantizer simple and predictable, do any decomps needed to get correct qparams in transform_for_annotation, and then do fusings in the cortex-m lowering pass pipeline, if this is possible rather than the derived-qparam approach. Having a clear split of responsibilities across the different stages of the lowering makes it easier to understand.

**Comment by amacharla15** [comment]
Thanks @AdrianLundell , I implemented the `transform_for_annotation` approach you suggested as #22892.

It moves the split default-SDPA score scale after the score BMM, keeps `CortexMQuantizer` unchanged, and leaves downstream fusion separate in the lowering pipeline. With the stock quantizer, all three attention score scales in the Transformer reproducer now get the proportional qparams needed for a later fold; I also checked this across head dimensions 16/32/64/128.

I’ve kept the linear folded-Q/DQ issue separate as discussed.
